### PR TITLE
[FLINK-33027] Allow users to override parallelism of excluded vertices

### PR DESCRIPTION
**Pull request by gyfora**
## What is the purpose of the change

1. Move parallelism override application to JobAutoscaler instead of reconciler to not leak autoscaler logic
2. Allow users to override excluded vertex parallelisms to avoid the corner case where excluded vertex parallelisms are stuck

## Verifying this change

Improved and extended unit test coverage

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-kubernetes-operator-autoscaler/src/test/java/org/apache/flink/kubernetes/operator/autoscaler/JobAutoScalerImplTest.java [hunk 1]
```diff
@@ -196,4 +199,75 @@ void testParallelismOverrides() throws Exception {
...
+                .put(PipelineOptions.PARALLELISM_OVERRIDES.key(), v1 + ":1");
+        autoscaler.applyParallelismOverrides(getResourceContext(deployment));
+        assertEquals(specClone, deployment.getSpec());
+
+        // Define partly overlapping overrides
+        var v2 = new JobVertexID();
+        deployment
+                .getSpec()
+                .getFlinkConfiguration()
+                .put(PipelineOptions.PARALLELISM_OVERRIDES.key(), v1 + ":1," + v2 + ":4");
```
Comment (gaborgsomogyi) [inline flink-kubernetes-operator-autoscaler/src/test/java/org/apache/flink/kubernetes/operator/autoscaler/JobAutoScalerImplTest.java:249]: OFFTOPIC: As a general user it would be good to see in the doc how to define maps in the config.

File: flink-kubernetes-operator-autoscaler/src/test/java/org/apache/flink/kubernetes/operator/autoscaler/JobAutoScalerImplTest.java [hunk 2]
```diff
@@ -196,4 +199,75 @@ void testParallelismOverrides() throws Exception {
...
+        assertEquals(
+                Map.of(v1.toHexString(), "2"),
+                getResourceContext(deployment)
+                        .getDeployConfig(deployment.getSpec())
+                        .get(PipelineOptions.PARALLELISM_OVERRIDES));
+
+        // Test when there are also user overrides, autoscaler should take precedence
+        specClone = ReconciliationUtils.clone(deployment.getSpec());
+
+        // This should be ignored
```
Comment (gaborgsomogyi) [inline flink-kubernetes-operator-autoscaler/src/test/java/org/apache/flink/kubernetes/operator/autoscaler/JobAutoScalerImplTest.java:236]: It would be good to add a minimal explanation why: `, because ...`

**Review by gaborgsomogyi — APPROVED** [review]
LGTM. Some explanation would be good to speed up dev's/user's understanding but that's optional.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/JobAutoScaler.java [hunk 3]
```diff
@@ -30,6 +28,6 @@ public interface JobAutoScaler {
     /** Called when the custom resource is deleted. */
     void cleanup(FlinkResourceContext<?> ctx);
 
-    /** Get the current parallelism overrides for the job. */
-    Map<String, String> getParallelismOverrides(FlinkResourceContext<?> ctx);
+    /** Apply current parallelism overrides. */
+    void applyParallelismOverrides(FlinkResourceContext<?> ctx);
```
Comment (mxm) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/JobAutoScaler.java:32]: Is this interface method really necessary? Overrides should be applied in the scale method, no?

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/JobAutoScaler.java:32]: This is based on earlier discussion in https://issues.apache.org/jira/browse/FLINK-32589 
Overrides need to be applied separately (and it was implemented already this way I just relocated the logic) so that we do not lose them on spec changes.

Comment (mxm) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/JobAutoScaler.java:32]: I couldn't find any discussion in that thread which would explain why the overrides couldn't be applied without the method here. Perhaps the code would need refactoring (which is done here already) to apply the scale method before reconciling other parts of the spec.

**Review by mxm — APPROVED** [review]
Thanks for the cleanup! Ideally, autoscaling would manage the overrides end-to-end but this is already cleaner than before.

**Comment by gyfora** [comment]
> Thanks for the cleanup! Ideally, autoscaling would manage the overrides end-to-end but this is already cleaner than before.

I agree, but for that as you said we need to first call the scale method before anything else. Currently we do the opposite and call scale after everything else. We have to think this through to make sure we don't introduce some problem in the general reconciler / autoscaler logic
